core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/internal/CoreRemoteMongoClientImpl.java

```diff
@@ -16,39 +16,55 @@
 
 package com.mongodb.stitch.core.services.mongodb.remote.internal;
 
-import com.mongodb.client.MongoClient;
-import com.mongodb.client.MongoDatabase;
-import com.mongodb.stitch.core.internal.common.AuthMonitor;
-import com.mongodb.stitch.core.internal.common.Dispatcher;
-import com.mongodb.stitch.core.internal.net.NetworkMonitor;
+import com.mongodb.stitch.core.StitchAppClientInfo;
 import com.mongodb.stitch.core.services.internal.CoreStitchServiceClient;
+import com.mongodb.stitch.core.services.internal.StitchServiceBinder;
+import com.mongodb.stitch.core.services.mongodb.local.internal.EmbeddedMongoClientFactory;
 import com.mongodb.stitch.core.services.mongodb.remote.sync.internal.DataSynchronizer;
+import com.mongodb.stitch.core.services.mongodb.remote.sync.internal.SyncMongoClientFactory;
 
-public class CoreRemoteMongoClientImpl implements CoreRemoteMongoClient {
+public class CoreRemoteMongoClientImpl implements CoreRemoteMongoClient, StitchServiceBinder {
 
   private final CoreStitchServiceClient service;
-  private final DataSynchronizer dataSynchronizer;
-  private final NetworkMonitor networkMonitor;
-  private final MongoDatabase tempDb;
+  private final StitchAppClientInfo appInfo;
+  private final EmbeddedMongoClientFactory clientFactory;
+
+  private DataSynchronizer dataSynchronizer;
 
   public CoreRemoteMongoClientImpl(final CoreStitchServiceClient service,
                                    final String instanceKey,
-                                   final MongoClient localClient,
-                                   final NetworkMonitor networkMonitor,
-                                   final AuthMonitor authMonitor,
-                                   final Dispatcher eventDispatcher) {
+                                   final StitchAppClientInfo appInfo,
+                                   final EmbeddedMongoClientFactory clientFactory) {
     this.service = service;
-    this.networkMonitor = networkMonitor;
-    this.tempDb = localClient.getDatabase("sync_temp");
+    this.appInfo = appInfo;
+    this.clientFactory = clientFactory;
+    this.service.bind(this);
 
     this.dataSynchronizer = new DataSynchronizer(
         instanceKey,
         service,
-        localClient,
+        SyncMongoClientFactory.getClient(
+            appInfo,
+            service.getName(),
+            clientFactory
+        ),
         this,
-        networkMonitor,
-        authMonitor,
-        eventDispatcher
+        appInfo.getNetworkMonitor(),
+        appInfo.getAuthMonitor(),
+        appInfo.getEventDispatcher()
+    );
+  }
+
+  @Override
+  public void onRebindEvent() {
+    // reinitialize the DataSynchronizer entirely.
+    // any auth event will trigger this.
```

Comment:
[minor] the act of adding an auth listener triggers an auth event, meaning this will always be called when starting an app for the first time. I could see that being a problem.

We could refactor the rebinding to only trigger when the user has actually changed, but I could imagine a scenario in the future where an auth event that didn't change the user actually would warrant a service rebinding. 

Maybe it would be best in `DataSynchronizer#reinitialize` to use the `AuthMonitor` to check if the active user id has changed before actually reinitializing. The DataSynchronizer can store the active user whose data is being synchronized.

Reply:
Good catch. I think it makes more sense to do it here, in the RemoteMongoClient.